xdsl/interpreter.py

```
@@ -0,0 +1,131 @@
+        yield self
+
+    def stack_description(self) -> str:
+        return '/'.join(c.name for c in self.stack())
+
+
+@dataclass
+class Intepreter:
+
+    module: ModuleOp
+    _function_table: Functions = field(default_factory=Functions)
+    _env: IntepretationEnv = field(
+        default_factory=lambda: IntepretationEnv(name='root'))
+    file: StringIO | None = field(default=None)
+
+    def get_values(self, values: tuple[SSAValue, ...]) -> tuple[Any, ...]:
+        return tuple(self._env[value] for value in values)
+
+    def set_values(self, ssa_values: tuple[SSAValue, ...],
+                   result_values: tuple[Any, ...]):
+        self._assert(
+            len(ssa_values) == len(result_values),
+            f'{[f"{ssa_value}" for ssa_value in ssa_values]}, {result_values}')
+        for ssa_value, result_value in zip(ssa_values, result_values):
+            self._env[ssa_value] = result_value
+
+    def push_scope(self, name: str = 'child') -> None:
+        self._env = IntepretationEnv(name, self._env)
+
+    def pop_scope(self) -> None:
+        if self._env.parent is None:
+            raise IntepretationError('Attempting to pop root env')
+
+        self._env = self._env.parent
+
+    def register_functions(self, funcs: Functions) -> None:
+        self._function_table.register_from(funcs)
+
+    def run(self, op: Operation):
+        op_type = type(op)
+        if op_type not in self._function_table.functions:
+            raise IntepretationError(
+                f'Could not find intepretation function for op {op.name}')
+
+        inputs = self.get_values(op.operands)
+        results = self._function_table.run(self, op, inputs)
+        self.set_values(tuple(op.results), results)
+
+    def print(self, *args: Any, **kwargs: Any):
+        print(*args, **kwargs, file=self.file)
+
+    def _assert(self, condition: bool, message: str | None = None):
+        assert condition, f'({self._env.stack_description()})({message})'
```

Maybe raise an Exception instead? This is something that I would consider API, so Exception raising is the way to got :)

---

- [ ] raise exception instead